### PR TITLE
Drop raft node

### PR DESCRIPTION
**Pull request by corylanou**
Currently, if you drop a raft node (even the leader), to bring another raft node to the cluster, you must launch a NEW node.  Shutting down an existing node and bringing it back up does not make it a raft by default.  This will be fixed in a future PR.

### CODE REVIEW

**Comment by otoolep** [comment]
What happens if the node being dropped is in communication with the cluster?

**Comment by otoolep** [comment]
Working with @corylanou here, seeing some funnies on non-dropped nodes after dropping the node.

File: meta/store.go [hunk 1]
```diff
@@ -519,7 +535,9 @@ func (s *Store) createLocalNode() error {
 	}
 
 	// Set ID locally.
+	s.mu.Lock()
```
Comment (otoolep) [inline meta/store.go]: Can you add a comment here, outlining why this locking is done?

**Comment by otoolep** [comment]
After more system debug, looks like a false alarm. @corylanou doing more system-level testing on systems with a dropped node, even if the node was up and running when dropped.

File: meta/store.go (hunk 1 above)
Comment (corylanou) [inline meta/store.go]: This was detected as a race by our race detector.  Likely only due to test code and would not be a race condition in production, but better safe than sorry.

Comment (otoolep) [inline meta/store.go]: I meant a comment in the code. :-)

File: CHANGELOG.md [hunk 2]
```diff
@@ -21,6 +21,7 @@
 - [#4375](https://github.com/influxdb/influxdb/pull/4375): Add Subscriptions so data can be 'forked' out of InfluxDB to another third party.
 - [#4506](https://github.com/influxdb/influxdb/pull/4506): Register with Enterprise service and upload stats, if token is available.
 - [#4501](https://github.com/influxdb/influxdb/pull/4501): Allow filtering SHOW MEASUREMENTS by regex.
+- [#4547](https://github.com/influxdb/influxdb/pull/4547): Drop raft node.
```
Comment (otoolep) [inline CHANGELOG.md]: Just so the guys reading the CHANGELOG know what this about is about can you rephrase it as something like:

`DROP SERVER` for all node types (or something).

File: meta/state.go [hunk 3]
```diff
@@ -318,6 +319,18 @@ func (r *localRaft) addPeer(addr string) error {
 	return nil
 }
 
+// removePeer removes addr from the list of peers in the cluster.
+func (r *localRaft) removePeer(addr string) error {
+	// Only do this on the leader
```
Comment (otoolep) [inline meta/state.go]: Should this function return an error like `NotLeader` if not the actual leader?

File: meta/statement_executor_test.go [hunk 4]
```diff
@@ -166,8 +166,12 @@ func TestStatementExecutor_ExecuteStatement_DropServer(t *testing.T) {
 		}, nil
 	}
 
-	// Ensure Raft nodes cannot be dropped.
-	if res := e.ExecuteStatement(influxql.MustParseStatement(`DROP SERVER 1`)); res.Err != meta.ErrNodeRaft {
+	e.Store.DeleteNodeFn = func(id uint64, force bool) error {
+		return nil
+	}
+
+	// Ensure Raft nodes can be dropped.
```
Comment (otoolep) [inline meta/statement_executor_test.go]: Minor: should this comment should actually read "ensure node can be dropped". It's not just Raft nodes, right?

File: meta/store.go [hunk 5]
```diff
@@ -454,15 +457,28 @@ func (s *Store) close() error {
...
+		s.Logger.Printf("error closing ExecListener %s", err)
+	}
 
 	if s.raftState != nil {
 		s.raftState.close()
-		s.raftState = nil
 	}
 
+	// Notify goroutines of close.
+	s.mu.Unlock()
```
Comment (otoolep) [inline meta/store.go]: Actually, I think this is where we could do with a comment about why we unlock here.

File: meta/store.go [hunk 6]
```diff
@@ -1705,6 +1746,33 @@ func (fsm *storeFSM) Apply(l *raft.Log) interface{} {
...
+func (fsm *storeFSM) applyRemovePeerCommand(cmd *internal.Command) interface{} {
+	ext, _ := proto.GetExtension(cmd, internal.E_RemovePeerCommand_Command)
+	v := ext.(*internal.RemovePeerCommand)
+
+	id := v.GetID()
+	addr := v.GetAddr()
+
+	// Only do this if you are the leader
+	if fsm.raftState.isLeader() {
+		//Remove that node from the peer
```
Comment (otoolep) [inline meta/store.go]: Nit: whitespace before "Remote".

**Comment by otoolep** [comment]
Some feedback, but this generally all makes sense to me. +1

File: meta/store.go (hunk 6 above)
Comment (corylanou) [inline meta/store.go]: Not sure what causes this.  It's fine locally, and go fmt doesn't see it either.  I even opened it in a couple of other editors to be sure as well.  Apparently only github sees the extra space?  I've seen this before as well and I'm still not sure what is going on.

File: meta/statement_executor_test.go (hunk 4 above)
Comment (corylanou) [inline meta/statement_executor_test.go]: no, this is testing specifically that a node that is a raft node can be dropped, and the next test drops a non-raft node.

File: meta/state.go (hunk 3 above)
Comment (corylanou) [inline meta/state.go]: fixed
